### DISCUSSION TITLE
Unguided Exercise - Express.js - Advanced Virtual Science Lab Exercise: Titration Simulation API - Part 1/3

### DISCUSSION THREAD

**Discussion by akash-coded** [akash-coded/mern#221] [Tasks]
### Scenario: Virtual Chemistry Lab

You are tasked with developing the backend for a virtual chemistry lab that simulates titration experiments. Students will be able to perform virtual titrations, adjust various parameters in real-time, and observe results, which will later be visualized using p5.js, d3.js, or three.js.

### Objective

Create a complex set of endpoints that simulate a titration experiment's backend, allowing the user to adjust concentrations, volumes, and to add indicators, with the system providing real-time feedback on the titration curve and equivalence point.

### Starter Data Structure

Create a new `titrationData.js` file:

```javascript
// titrationData.js

const acids = [
  // Array of common acids with properties like name, formula, and molarMass.
];

const bases = [
  // Array of common bases.
];

const indicators = [
  // Array of pH indicators with properties like name, transitionpHRange, and colors.
];

const titrations = [
  // Record of titrations with unique identifiers and their parameters.
];

module.exports = { acids, bases, indicators, titrations };
```

### Server Setup and Endpoint Definition

1. **Initialize the project and install Express**:

   - Set up the Node project with `npm init`.
   - Install Express with `npm install express`.

2. **Create a `server.js` file**:

   - Import the starter data from `titrationData.js`.
   - Set up middleware (body-parser, CORS).

3. **Define Your Endpoints**:

   - `/acids` and `/bases` - To retrieve and update the list of acids and bases.
   - `/indicators` - To manage pH indicators.
   - `/titrations` - To conduct titration experiments.

### CRUD Operation Endpoints

**Acids and Bases Management**

- GET all acids/bases
- GET an acid/base by ID
- POST a new acid/base
- PUT to update an acid/base by ID
- DELETE an acid/base by ID

**Indicator Management**

- GET all indicators
- GET an indicator by ID
- POST a new indicator
- PUT to update an indicator by ID
- DELETE an indicator by ID

**Titration Experiment Management**

- GET all titrations
- GET a titration by ID
- POST to start a new titration
- PUT to update titration parameters by ID
- DELETE a titration record by ID

### Testing Endpoints with Postman

Define Postman requests for each type of operation:

- **Acids/Bases and Indicator CRUD**:
  - Create, read, update, and delete various chemical solutions and indicators.
- **Titration Experiment Control**:
  - Start, adjust, and delete titration experiments.

### Sample Payload for Postman Requests

```json
// POST /acids
{
  "name": "Hydrochloric Acid",
  "formula": "HCl",
  "molarMass": 36.46
}

// PUT /titrations/:id
{
  "volumeAdded": 5,
  "concentration": 0.1
}
```

### Considerations

- Implement input validation and error handling.
- Design a formula to calculate pH and the titration curve based on user input.
- Consider how your API will scale once connected to a MongoDB database.
- Plan how the API will communicate with the frontend visualization tools.

### Advanced Challenges

- Simulate real-time data streaming with `socket.io` for live titration curve updates.
- Introduce user sessions with authentication to save and retrieve past experiments.
- Create an algorithm that automatically identifies the equivalence point based on the titration data.

### Wrapping Up

This exercise demands a comprehensive approach to simulate a chemistry lab environment, creating an intricate system that manages a variety of entities and processes complex calculations. By developing this API, you will lay the groundwork for a sophisticated educational tool that brings the principles of chemistry to life through interactive simulations.

**Comment by akash-coded** [comment]
### Sample Data

To create an extensive set of dummy hard-coded data for a titration simulation application, we will define several arrays for acids, bases, indicators, and titrations. This data will mimic what you might expect in a real-world chemistry database.

### acids.js
```javascript
const acids = [
  {
    id: 1,
    name: "Hydrochloric Acid",
    formula: "HCl",
    molarMass: 36.46
  },
  {
    id: 2,
    name: "Sulfuric Acid",
    formula: "H2SO4",
    molarMass: 98.079
  },
  {
    id: 3,
    name: "Nitric Acid",
    formula: "HNO3",
    molarMass: 63.012
  },
  // ... more acid data
];

module.exports = acids;
```

### bases.js
```javascript
const bases = [
  {
    id: 1,
    name: "Sodium Hydroxide",
    formula: "NaOH",
    molarMass: 40.00
  },
  {
    id: 2,
    name: "Potassium Hydroxide",
    formula: "KOH",
    molarMass: 56.11
  },
  {
    id: 3,
    name: "Calcium Hydroxide",
    formula: "Ca(OH)2",
    molarMass: 74.09
  },
  // ... more base data
];

module.exports = bases;
```

### indicators.js
```javascript
const indicators = [
  {
    id: 1,
    name: "Litmus",
    transitionpHRange: { acidic: 6.5, basic: 8.3 },
    colors: { acidic: "red", neutral: "purple", basic: "blue" }
  },
  {
    id: 2,
    name: "Phenolphthalein",
    transitionpHRange: { acidic: 8.2, basic: 10.0 },
    colors: { acidic: "colorless", neutral: "colorless", basic: "pink" }
  },
  {
    id: 3,
    name: "Methyl Orange",
    transitionpHRange: { acidic: 3.1, basic: 4.4 },
    colors: { acidic: "red", neutral: "orange", basic: "yellow" }
  },
  // ... more indicator data
];

module.exports = indicators;
```

### titrations.js
```javascript
const titrations = [
  {
    id: 1,
    acid: {
      name: "Hydrochloric Acid",
      concentration: 0.1 // Molarity
    },
    base: {
      name: "Sodium Hydroxide",
      concentration: 0.1 // Molarity
    },
    indicator: {
      name: "Phenolphthalein",
      transitionpHRange: { acidic: 8.2, basic: 10.0 }
    },
    volumeBaseAdded: 0, // in mL
    pH: 7, // initial pH value for simulation
    observations: []
    // ... this array will be populated with observation data during the titration simulation
  },
  // ... more titration records
];

module.exports = titrations;
```

This data can be used to simulate the behavior of a titration experiment in the virtual lab. When creating the simulation logic, you'll use these data arrays to calculate the pH at each step of the titration, determine the color of the solution based on the indicator used, and identify the equivalence point.

The `observations` array within each titration would be populated with data points representing the volume of base added and the corresponding pH, creating a titration curve that can be graphed in the frontend. This approach lays the groundwork for a rich and interactive educational experience without the need for an actual database, which could be added later to enhance the application's features.

**Comment by akash-coded** [comment]
Building upon the detailed dummy data provided previously, let’s continue and populate the `titrations` data structure with more entries and observation data to give a rich dataset for initial development and testing.

### Expanded titrations.js

We will simulate a few titration experiments, each with a series of observations that would represent data collected at various stages of the titration process.

```javascript
const titrations = [
  {
    id: 1,
    acid: {
      name: "Hydrochloric Acid",
      concentration: 0.1 // Molarity
    },
    base: {
      name: "Sodium Hydroxide",
      concentration: 0.1 // Molarity
    },
    indicator: {
      name: "Phenolphthalein",
      transitionpHRange: { acidic: 8.2, basic: 10.0 }
    },
    volumeBaseAdded: 50, // in mL
    pH: 7, // initial pH value for simulation
    observations: [
      { volumeBaseAdded: 0, pH: 2 },
      { volumeBaseAdded: 5, pH: 2.5 },
      { volumeBaseAdded: 10, pH: 3.1 },
      { volumeBaseAdded: 15, pH: 4.5 },
      // ...
      // More observation data
      { volumeBaseAdded: 50, pH: 11 }
    ]
  },
  {
    id: 2,
    acid: {
      name: "Acetic Acid",
      concentration: 0.1 // Molarity
    },
    base: {
      name: "Potassium Hydroxide",
      concentration: 0.1 // Molarity
    },
    indicator: {
      name: "Litmus",
      transitionpHRange: { acidic: 6.5, basic: 8.3 }
    },
    volumeBaseAdded: 30, // in mL
    pH: 7, // initial pH value for simulation
    observations: [
      { volumeBaseAdded: 0, pH: 3 },
      { volumeBaseAdded: 5, pH: 4 },
      { volumeBaseAdded: 10, pH: 4.7 },
      // ...
      // More observation data
      { volumeBaseAdded: 30, pH: 8 }
    ]
  },
  // ... additional titration records with different acids, bases, and indicators
];

module.exports = titrations;
```

This `titrations` array will serve as a pivotal dataset for the backend services to simulate a virtual titration lab. Each titration contains a complete set of observations that mimic the actual data points one would collect during a laboratory titration. This includes the incremental addition of base into the acid solution and the corresponding pH changes.

The observation data should ideally cover the entire range of the titration process, from the initial acidic pH to the final basic pH, passing through the neutralization point where the indicator changes color.

For the initial development, the focus should be on creating a robust set of API endpoints to perform CRUD operations on these datasets. You’ll need to ensure that you can:

- Retrieve the entire list of titrations or a single titration by ID.
- Add a new titration with initial parameters for the acid, base, and indicator.
- Update an existing titration with new observation data (such as adding a new observation for a volume of base added and the new pH).
- Delete a titration record.

This structured data will be invaluable for testing the API, ensuring each endpoint behaves as expected, and later, for integrating the backend with the frontend visualization tools, which will render the titration curve and allow for interactive simulations.

**Comment by UmarAikawa** [comment]
const express = require("express");
const bodyParser = require('body-parser');
// const titrationData = require("./titrationData.js");
const cors = require('cors');

const app = express();
const port = 3000;

app.use(express.json());
app.use(cors());
app.use(bodyParser.json());
app.use(bodyParser.urlencoded({ extended: true }));


app.listen(port, () => {
    console.log(`Microscope server listening at http://localhost:${port}`);
});

// mongoose
const mongoose = require('mongoose');
// connect to the database
mongoose.connect('mongodb://localhost/titrationDB', { useNewUrlParser: true, useUnifiedTopology: true });

// Define an acid schema
const acidSchema = new mongoose.Schema({
  id: Number,
  name: String,
  formula: String,
  molarMass: Number,
});

// Create an acid model
const Acid = mongoose.model('Acid', acidSchema);

// end points

// show all acids 
app.get('/acids', (req, res) => {
    res.json({ acids: Acid });
});

// create an acid
app.post('/acid', async (req, res) => {
    try {
        let acid = new Acid(req.body);
        acid = await acid.save();
        res.status(201).send(acid);
    } catch (error) {
        res.status(400).send(error);
    }
});

// acids update to update list of acids
app.post('/acids', (req, res) => {
    let id = req.body.id;
    let name = req.body.name;
    // do update here
    res.send(`Acid no ${id} has been updated to ${name}`);
});

**Reply by UmarAikawa** [reply]
This is what I have achieved so far on this

**Reply by UmarAikawa** [reply]
//This is the completed code

const express = require("express");
const bodyParser = require('body-parser');
const cors = require('cors');

const app = express();
const port = 3000;

app.use(express.json());
app.use(cors());
app.use(bodyParser.json());
app.use(bodyParser.urlencoded({ extended: true }));


app.listen(port, () => {
    console.log(`Titaration server listening at http://localhost:${port}`);
});

// mongoose
const mongoose = require('mongoose');
// connect to the database
mongoose.connect('mongodb://127.0.0.1:27017/titrationDB', { useNewUrlParser: true, useUnifiedTopology: true });

// Define an acid schema
const acidSchema = new mongoose.Schema({
  id: Number,
  name: String,
  formula: String,
  molarMass: Number,
});

// Create base model
const Acid = mongoose.model('Acid', acidSchema);

// Define base schema
const baseSchema = new mongoose.Schema({
    id: Number,
    name: String,
    formula: String,
    molarMass: Number,
});
    
// Create base model
const Base = mongoose.model('Base', baseSchema);

// Define an indicator schema
const indicatorSchema = new mongoose.Schema({
    id: Number,
    name: String,
    transitionpHRange: Object,
    colors: Object,     
});
    
// Create an indicator model
const Indicator = mongoose.model('Indicator', indicatorSchema);

// Define a titration schema
const titrationSchema = new mongoose.Schema({
    id: Number,
    acid: Object,
    base: Object,
    indicator: Object,
    volumeBaseAdded: Number,
    pH: Number,
    observations: Array,
});
    
// Create a titration model
const Titaration = mongoose.model('Titaration', titrationSchema);

// end points
// ACID
// show all acids 
app.get('/showacids', (req, res) => {
    Acid.find()
    .then(data => {
        res.send(data);
    }).catch(err => {
        res.status(500).send({
            message:
            err.message || "Some error occurred while retrieving acids."
        });
    });
});

// create an acid
app.post('/createacid', async (req, res) => {
    try {
        let acid = new Acid(req.body);
        acid = await acid.save();
        res.status(201).send(acid);
    } catch (error) {
        res.status(400).send(error);
    }
});

// acids update to update list of acids
app.post('/updateacid', (req, res) => {
    let acidId = req.body.id;
    let name = req.body.name;
    // do update here
    Acid.findOneAndUpdate({id: acidId},{$set:{name: name}})
    .then(data => {
        if (!data) {
        res.status(404).send({
            message: `Cannot update Acid with id=${acidId}. Maybe Acid was not found!`
        });
        } else {
        res.send({
            message: "Acid was updated successfully!"
        });
        }
    })
    .catch(err => {
        res.status(500).send({
        message: "Could not update Acid with id=" + acidId
        });
    }); 
});

// delete acid
app.post('/deleteacid', (req, res) => {
    let acidId = req.body.id;
    Acid.deleteOne({ id: acidId })
    .then(data => {
        if (!data) {
        res.status(404).send({
            message: `Cannot delete Acid with id=${acidId}. Maybe Acid was not found!`
        });
        } else {
        res.send({
            message: "Acid was deleted successfully!"
        });
        }
    })
    .catch(err => {
        res.status(500).send({
        message: "Could not delete Acid with id=" + acidId
        });
    });
});
// BASE
// show all bases 
app.get('/showbases', (req, res) => {
    Base.find()
    .then(data => {
        res.send(data);
    }).catch(err => {
        res.status(500).send({
            message:
            err.message || "Some error occurred while retrieving bases."
        });
    });
});

// create a base
app.post('/createbase', async (req, res) => {
    try {
        let base = new Base(req.body);
        base = await base.save();
        res.status(201).send(base);
    } catch (error) {
        res.status(400).send(error);
    }
});

// bases update to update list of bases
app.post('/updatebase', (req, res) => {
    let baseId = req.body.id;
    let name = req.body.name;
    // do update here
    Base.findOneAndUpdate({name: name},{$set:{id: baseId}})
    .then(data => {
        if (!data) {
        res.status(404).send({
            message: `Cannot update Base with id=${baseId}. Maybe Base was not found!`
        });
        } else {
        res.send({
            message: "Base was updated successfully!"
        });
        }
    })
    .catch(err => {
        res.status(500).send({
        message: "Could not update Base with id=" + baseId
        });
    }); 
});

// delete base
app.post('/deletebase', (req, res) => {
    let baseId = req.body.id;
    Base.deleteOne({ id: baseId })
    .then(data => {
        if (!data) {
        res.status(404).send({
            message: `Cannot delete Acid with id=${baseId}. Maybe Acid was not found!`
        });
        } else {
        res.send({
            message: "Acid was deleted successfully!"
        });
        }
    })
    .catch(err => {
        res.status(500).send({
        message: "Could not delete Acid with id=" + baseId
        });
    });
});

// INDICATORS
// show all indicators
app.get('/showindicators', (req, res) => {
    Indicator.find()
    .then(data => {
        res.send(data);
    }).catch(err => {
        res.status(500).send({
            message:
            err.message || "Some error occurred while retrieving indicators."
        });
    });
});

// create a indicator
app.post('/createindicator', async (req, res) => {
    try {
        let indicator = new Indicator(req.body);
        indicator = await indicator.save();
        res.status(201).send(indicator);
    } catch (error) {
        res.status(400).send(error);
    }
});

// indicator update to update list of indicators
app.post('/updateindicator', (req, res) => {
    let indicatorId = req.body.id;
    let name = req.body.name;
    // do update here
    Indicator.findOneAndUpdate({id: indicatorId},{$set:{name: name}})
    .then(data => {
        if (!data) {
        res.status(404).send({
            message: `Cannot update Indicator with id=${indicatorId}. Maybe Indicator was not found!`
        });
        } else {
        res.send({
            message: "Indicator was updated successfully!"
        });
        }
    })
    .catch(err => {
        res.status(500).send({
        message: "Could not update Indicator with id=" + indicatorId
        });
    }); 
});

// delete base
app.post('/deleteindicator', (req, res) => {
    let indicatorId = req.body.id;
    Indicator.deleteOne({ id: indicatorId })
    .then(data => {
        if (!data) {
        res.status(404).send({
            message: `Cannot delete Indicator with id=${indicatorId}. Maybe Indicator was not found!`
        });
        } else {
        res.send({
            message: "Indicator was deleted successfully!"
        });
        }
    })
    .catch(err => {
        res.status(500).send({
        message: "Could not delete Indicator with id=" + indicatorId
        });
    });
});

// TITRATION
// show all titrations
app.get('/showtitrations', (req, res) => {
    Titaration.find()
    .then(data => {
        res.send(data);
    }).catch(err => {
        res.status(500).send({
            message:
            err.message || "Some error occurred while retrieving titrations."
        });
    });
});

// create a titration
app.post('/createtitration', async (req, res) => {
    try {
        let titration = new Titaration(req.body);
        titration = await titration.save();
        res.status(201).send(titration);
    } catch (error) {
        res.status(400).send(error);
    }
});

// indicator update to update list of titrations
app.put('/updatetitration', (req, res) => {
    let titrationId = req.body.id;
    let observs = req.body.observations;
    // do update here
    Titaration.findOneAndUpdate({id: titrationId},{$set:{observations: observs}})
    .then(data => {
        if (!data) {
        res.status(404).send({
            message: `Cannot update Titration with id=${titrationId}. Maybe Titration was not found!`
        });
        } else {
        res.send({
            message: "Titration was updated successfully!"
        });
        }
    })
    .catch(err => {
        res.status(500).send({
        message: "Could not update Titration with id=" + titrationId
        });
    }); 
});

// delete titration
app.post('/deletetitration', (req, res) => {
    let titrationId = req.body.id;
    Titaration.deleteOne({ id: titrationId })
    .then(data => {
        if (!data) {
        res.status(404).send({
            message: `Cannot delete Titration with id=${titrationId}. Maybe Titration was not found!`
        });
        } else {
        res.send({
            message: "Titration was deleted successfully!"
        });
        }
    })
    .catch(err => {
        res.status(500).send({
        message: "Could not delete Titration with id=" + titrationId
        });
    });
});
